File: grant_clean/100.py
# cnki parser
import re
import os
import csv
import time
import datetime
import configparser
from bs4 import BeautifulSoup
from multiprocessing import Pool, cpu_count

from db import sql_execute, sql_insert, sql_insert_many
from tools import write_to_csv

def cleaner(row) :
    features = {
        "id"                : row["id"],
        "grant_sites_id"    : row["grant_sites_id"],
        "project_id"        : None,
        "project_id_original" : None,
        "title"             : None,
        "title_en"          : None,
        "project_url"       : row["project_url"],
        "keyword"           : None,
        "keyword_en"        : None,
        "source"            : None,
        "researcher"        : None,
        "researcher_en"     : None,
        "researcher_cleaned" : None,
        "institution"       : None,
        "institution_en"       : None,
        "institution_cleaned" : None,
        "country"           : "DE",
        "currency"          : None,
        "award_amount"      : None,
        "award_amount_usd"  : None,
        "funded_date"       : None,
        "start_year"        : None,
        "end_year"          : None,
        "start_date"        : None,
        "end_date"          : None,
        "project_term"      : None,
        "description"       : None,
        "description_raw"   : None,
        "description_en"    : None,    
    }
    html = row["html"]
    del row

    if not html :
        # print("html not found")
        return features
    
    try :
        soup = BeautifulSoup(html, "html.parser")
    except :
        # print("Soup errors!")
        return features
    
    features["title"]    = get_title(soup)
    features["title_en"] = features["title"]

    features["project_id"]          = get_project_id(features["project_url"])
    features["project_id_original"] = features["project_id"]
    
    features["researcher"], features["institution"] = get_researcher(soup)

    features["description"]  = get_description(soup)
    features["description_raw"] = features["description"]

    features["start_year"], features["end_year"] = get_end_year(soup)

    return features

def get_researcher(soup) :
    try :
        spans = soup.find_all("span", {"class" : ["name", "value"]})
    except :
        return None, None

    researcher, institution = [], []
    for i in range(0, len(spans) if len(spans) % 2 == 0 else len(spans) - 1, 2) :
        cur_span = spans[i]
        next_span = spans[i + 1]

        cur_text = cur_span.get_text().replace("­", "")
        if cur_text.find("Antragsteller") != -1 :
            try :
                for row in next_span.find_all("a") :
                    researcher.append(row.get_text().strip())
            except :
                pass
            [s.extract() for s in next_span.find_all("a")]
            next_span = str(next_span).replace("<br/>", ";")
            soup = BeautifulSoup(next_span, "html.parser")
            try :
                for row in soup.get_text().split(";") :
                    if row.strip() != "" :
                        institution.append(row.strip())
            except :
                pass
        regex_string = r"Teilprojektleiter|Leiter|Wis­sen­schaft­le­rin­nen|Sprecher"
        if re.search(regex_string, cur_text) :
            try :
                for row in next_span.find_all("a") :
                    researcher.append(row.get_text().strip())
            except :
                pass

        if cur_text.find("Institution") != -1 :
            try :
                for row in next_span.find_all("a") :
                    institution.append(row.get_text().strip())
            except :
                pass
    researcher = ";".join(researcher) if researcher else None
    institution = ";".join(institution) if institution else None

    return researcher, institution

def get_title(soup) :
    try :
        return soup.find("div", {"class" : "details"}).find("h3").get_text().strip()
    except :
        return None
    return None

def get_description(soup) :
    try :
        return soup.find("div", {"id" : "projekttext"}).get_text().strip()
    except :
        return None
    return None

def get_end_year(soup) :
    try :
        spans = soup.find_all("span", {"class" : ["name", "value"]})
    except :
        return None
    start_year, end_year = None, None
    for i in range(0, len(spans) if len(spans) % 2 == 0 else len(spans) - 1, 2) :
        cur_span = spans[i]
        next_span = spans[i + 1]

        if cur_span.get_text().strip() == "Förderung" :
            text = next_span.get_text()
            regex_string = r"\b\d{4}\b"
            m = re.findall(regex_string, text)
            if m :
                if len(m) == 2 :
                    start_year = min(m[0], m[1])
                    end_year = max(m[0], m[1])
                elif len(m) == 1 :
                    start_year = m[0]

    return start_year, end_year

def get_project_id(project_url) :
    if project_url is None :
        return None
    project_id = project_url.split("/")[-1].strip()

    try :
        int(project_id)
    except :
        return None
    return project_id

def multi_process_db() :
    sql = """SELECT * FROM tmp_dfg_html WHERE html is not null"""
    query_result = sql_execute(sql, db_config)

    sub_grant_datas = []

    for row in query_result :
        features = {
            "id"              : row[0],
            "grant_sites_id"  : row[1],
            "project_url"     : row[2],
            "html"            : row[3],
        }
        sub_grant_datas.append(features)

    p = Pool(processes=cpu_count())
    grant_datas = p.map(cleaner, sub_grant_datas)
    p.close()
    p.join()

    del sub_grant_datas
    sql = """INSERT INTO tmp_dfg_info(
        id,
        grant_sites_id,
        project_id,
        project_id_original,
        title,
        title_en,
        project_url,
        keyword,
        keyword_en,
        source,
        researcher,
        researcher_en,
        researcher_cleaned,
        institution,
        institution_en,
        institution_cleaned,
        country,
        currency,
        award_amount,
        award_amount_usd,
        funded_date,
        start_year,
        end_year,
        start_date,
        end_date,
        project_term,
        description,
        description_raw,
        description_en,
        created_date,
        updated_date,
        crawled_date)
        VALUES(%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s, now(),now(),now() )"""
    
    input_datas = []
    
    for row in grant_datas :
        input_datas.append((
            int(row["id"]),
            int(row["grant_sites_id"]),
            row["project_id"],
            row["project_id_original"],
            row["title"],
            row["title_en"],
            row["project_url"],
            row["keyword"],
            row["keyword_en"],
            row["source"],
            row["researcher"],
            row["researcher_en"],
            row["researcher_cleaned"],
            row["institution"],
            row["institution_en"],
            row["institution_cleaned"],
            row["country"],
            row["currency"],
            row["award_amount"],
            row["award_amount_usd"],
            row["funded_date"],
            row["start_year"],
            row["end_year"],
            row["start_date"],
            row["end_date"],
            row["project_term"],
            row["description"],
            row["description_raw"],
            row["description_en"]
        ))

    sql_insert_many(sql, db_config, input_datas)

if __name__ == "__main__":
    start_time = time.time()

    config = configparser.ConfigParser()
    config.read('.env')

    db_config = (config["DEFAULT"]["_HOSTNAME"], config["DEFAULT"]["_DATABASE"], config["DEFAULT"]["_USERNAME"], \
    config["DEFAULT"]["_PASSWORD"], config["DEFAULT"]["_PORT"])

    multi_process_db()

    # row = {
    #     "id" : 1,
    #     "grant_sites_id" : 100,
    #     "project_url" : "http://gepris.dfg.de/gepris/projekt/73376775"
    # }
    # with open("./tmp/test.html") as f :
    #     row["html"] = f.read()

    # tmp = cleaner(row)

    # for key, val in tmp.items() :
    #     print(key, val)

    print("Total time is : ", str(time.time() - start_time))

File: grant_clean_js/100.js
const cheerio = require('cheerio');

function getResearcher(cheer) {
    var spans = [];
    try {
        cheer("span.name, span.value").each(function(i, elem){
            spans.push(cheer(this));
        });
    }
    catch(error) {
        return [null, null];
    }
    var researcher = [];
    var institution = [];
    //(i in range(0, len(spans) if len(spans) % 2 == 0 else len(spans) - 1, 2) ) 
    var len = spans.length%2 == 0 ? spans.length : spans.length-1;
    for(let i = 0; i <  len; i+=2){ 
        var curSpan = spans[i];
        var nextSpan = spans[i + 1];
        if(curSpan) {
            var curText = curSpan.text().replace(/­/g, "");
        }else {
            continue;
        }
        // if(curText.indexOf("Antragsteller") != -1||curText.indexOf('An­trag­stel­le­rin­nen') != -1 || curText.indexOf('An­trag­stel­ler') != -1 || curText.indexOf('Teil­pro­jekt­lei­te­rin­nen / Teil­pro­jekt­lei­ter') != -1 || curText.indexOf('Teil­pro­jekt­lei­ter') != -1 ) {
        if(curText.indexOf("Antragsteller") != -1 || curText.indexOf('Teil­pro­jekt­lei­ter') != -1 ) {
            try {
                cheer = cheer.load(nextSpan.html());
                cheer("a").each(function(i,elem){
                    researcher.push(cheer(this).text().trim());
                });
                // for (row in nextSpan.find_all("a")) {
                //     researcher.append(row.get_text().strip())
                // }
            }
            catch(error) {

            }
            // [s.extract() for s in nextSpan.find_all("a")]
            nextSpan = cheer.load(nextSpan.html());
            nextSpan("a").each(function(i, elem){
                nextSpan(this).remove();
            });

            tmpSpan = nextSpan.html().replace(/<br\/>/g, ";");
            cheer = cheer.load(tmpSpan);
            try {
                for(row of cheer.text().split(";") ){
                    if (row.trim() != "" ){
                        institution.push(row.trim());
                    }
                }
            }
            catch(error) {}
        }
        regexString = /Teilprojektleiter|Leiter|Wis­sen­schaft­le­rin­nen|Sprecher/g
        //re.search(regexString, curText) {}
        // curText = curText.replace(/­/g, "")

        if (curText.search(regexString) != -1) {
            try {
                cheer = cheer.load(nextSpan.html())
                cheer("a").each(function(i, elem){
                    researcher.push(cheer(this).text().trim());
                });
                // for(row in nextSpan.find_all("a")) {
                //     researcher.append(row.get_text().strip())
                // }
            }
            catch(error) {}
                
        }
        if (curText.search("Institution") != -1) {
            try {
                cheer = cheer.load(nextSpan.html());
                cheer("a").each(function(i, elem){
                    institution.push(cheer(this).text().trim());
                });
                // for row in nextSpan.find_all("a") {}
                //     institution.append(row.get_text().strip())
            }
            catch(error) {
            }
        }
    }
    if(researcher.length > 0){
        researcher = researcher.join(";");
    }
    else{
        researcher = null;
    }
    if(institution.length > 0){
        institution = institution.join(";");
    }
    else{
        institution = null;
    }
    return [researcher, institution];
}

function cleanHtml(html){
    return html.toString().replace(/\n|\t|\s+/g," ");
    //re.sub(r"\n|\t|\s+", " ", str(soup))
}


function getTitle(cheer) {
    try {
        if(cheer("div.details").find("h3").text().trim() == "")
            return null;
        return cheer("div.details").find("h3").text().trim();
        // return cheer.find("div", {"class" {} "details"}).find("h3").get_text().strip()
    }
    catch(error) {
        return null;
    }
}
function getDescription(cheer) {
    try {
        if(cheer("div#projekttext").text().trim() == "")
            return null;
        return cheer("div#projekttext").text().trim();
        // return cheer.find("div", {"id" {} "projekttext"}).get_text().strip()
    }
    catch(error) {
        return null;
    }
}

function getEndYear(cheer) {
    var spans = [];
    try {
        cheer("span.name, span.value").each(async function(i, elem){
            spans.push(cheer(this));
        })
    }
    catch(error) {
        return [null,null];
    }
    var startYear = null;
    var endYear = null;
    for(let i = 0; i < spans.length-1; i++){ 
        curSpan = spans[i];
        nextSpan = spans[i + 1];
        if (curSpan.text().trim() == "F&#xF6;rderung" || curSpan.text().trim() == "Förderung") {
            text = nextSpan.text();
            regexString = /\b\d{4}\b/g;
            var m = text.match(regexString);
            //m = re.findall(regexString, text)
            if (m != null) {
                if (m.length == 2) {
                    startYear = Math.min(m[0], m[1]);
                    endYear = Math.max(m[0], m[1]);
                }
                else if( m.length == 1 ){
                    startYear = m[0];
                }
            }
        }
    }
    return [startYear, endYear];   
}

function getProjectId(projectUrl) {
    if (projectUrl == null) {
        return null
    }
    var projectId = projectUrl.split("/")[projectUrl.split("/").length-1].trim();

    try {
        var pid = parseInt(projectId);
    }
    catch(error) {
        return null;
    }
    return projectId;
}

async function parse(features) {
    var html = features['dtl_html'];

    if(html == null){
        return features;
    }
    html = cleanHtml(html);
    var cheer = cheerio.load(html, {decodeEntities: false});
    features["title"]    = getTitle(cheer);
    features["title_en"] = features["title"];

    features["project_id"]          = getProjectId(features["project_url"]);
    features["project_id_original"] = features["project_id"];
    var res = getResearcher(cheer);
    features["researcher"] = res[0];
    features["institution"] = res[1];

    features["description"]  = getDescription(cheer);
    features["description_raw"] = features["description"];
    let years = getEndYear(cheer);
    features["start_year"] = years[0];
    features["end_year"] = years[1];
    features['country'] = 'DE';
    features['currency'] = 'EUR';
    if( features["start_year"] != null && features["end_year"] !=null ) {
        features["project_term"] = parseInt(features["end_year"]) - parseInt(features["start_year"]) + 1 + " years"
    }else{
        features["project_term"] = null
    };

    return features;
}

module.exports = {
    parse,
  };
